Core/Auras: Implement SPELL_AURA_397 & SPELL_AURA_398

Golrag opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Define SPELL_AURA_397 & SPELL_AURA_398. More info can be found in #24794
-  Remove GetPlayerPositionData virtual function in Battleground class
-  Save the players in a vector in the base battleground class.

**Target branch(es):** 3.3.5/master

-  3.3.5 maybe
- master

**Issues addressed:**

Closes #24794


**Tests performed:**
- Builds
- Went in game, queued for Eye of the Storm (specific), cast 34976 and then with debugger went through everything. Also removed the buff by clicking on it and went through the debugger again. Also checked map if the flag was showing up, it did.


**Known issues and TODO list:**

- Names might not be very blizzlike
- Some coding standards?

**Notes:**
It's been a while.
I went for a vector because the order of insertion is being kept in the packets. (first player who takes flag is the first one in the position).
Unsure about making a new struct for it... Maybe I should have just used the one from the WorldPackets and updated the position each time. Not sure if that's something thats acceptable tho (in case of updating packets etc might get a little more complex)

Edit: Using the packet class makes things a lot cleaner tho

Any feedback is welcome :)

<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Golrag committed:
Core/Auras: Implement SPELL_AURA_397 & SPELL_AURA_398

- Replace Battleground code with it

illfated commented:
Looks interesting. Should be interesting to test for anyone who uses TC master branch and also enjoys Battlegrounds.

Golrag committed:
Use packet class to store data.

This way we can still broadcast the last known position of the player if we cant find the player for some reason.

Golrag committed:
Codestyle changes

Golrag committed:
Rename function

illfated commented:
Curious error message in AppVeyor (warning treated as error): `'WorldPackets::Battleground::BattlegroundPlayerPosition *const ': unknown size`

Golrag commented:
Dont have that error locally, so might be a little tricky to resolve

Golrag committed:
Move implementation out of header file

Golrag committed:
Move implementation of adding and removing player positions to battleground class

Golrag commented:
Well.. something I could try is to include the BattlegroundPackets.h in Battleground.h but I would like to avoid that. Anyone has other suggestions?

Shauren commented on src/server/game/Battlegrounds/Battleground.cpp:
```diff
@@ -1669,6 +1678,21 @@ void Battleground::PSendMessageToAll(uint32 entry, ChatMsg msgType, Player const
+
+void Battleground::RemovePlayerPosition(ObjectGuid guid)
+{
+    auto const& itr = std::remove_if(_playerPositions.begin(), _playerPositions.end(), [guid](WorldPackets::Battleground::BattlegroundPlayerPosition const info)
```
Do not const& this, the type here is `std::vector<WorldPackets::Battleground::BattlegroundPlayerPosition>::iterator` - you are making a reference to temporary (gets destroyed after `;`)

Golrag committed:
Remove reference

Mesielecat commented:
I think that the 398 the misc value is directly the iconId, if you check the 121164 Orb of Power or 121164 Orb of Power or 121176 Orb of Power or 121177 Orb of Power, this are out of the arena slot

Mesielecat commented:
Reviewing sniff the value of arenaSlot is the same that the misc value, then this arenaSlot is confuse and need update the name of the variable, maybe flagSlot

Golrag commented:
In my early bfa sniff I can see that temple of kot mogu always has IconID 1 or 2 and that the misc value is the ArenaSlot. IconID depends on faction and reverses it with the corresponding aura effect. I wrote that down in the issue I linked

Golrag commented:
To fix the build on appveyor i could try having a vector of pointers or unique_ptrs, will play around with it tomorrow

Golrag committed:
Store positions as a shared_ptr

illfated commented:
> To fix the build on appveyor i could try having a vector of pointers or unique_ptrs, will play around with it tomorrow

Seems like your changes did the trick. Thank you, well done. 👍🏻

Shauren commented:
Why shared_ptr?

Golrag commented:
Had some compile issues with a unique_ptr :/ can try to look at it again or go with just a pointer

Shauren commented on src/server/game/Spells/Auras/SpellAuraDefines.h:
```diff
@@ -469,8 +469,8 @@ enum AuraType : uint32
-    SPELL_AURA_397                                          = 397,
-    SPELL_AURA_398                                          = 398,
+    SPELL_AURA_UPDATE_BATTLEGROUND_PLAYER_POSITION          = 397,
+    SPELL_AURA_UPDATE_BATTLEGROUND_PLAYER_POSITION_2        = 398,
```
Proposed names
SPELL_AURA_BATTLEGROUND_PLAYER_POSITION_FACTIONAL = 397 (inspired by BATTLEMASTER_LIST_FLAG_FACTIONAL)
SPELL_AURA_BATTLEGROUND_PLAYER_POSITION = 398

Golrag replied:
done!

Golrag committed:
Rename aura types

Golrag commented:
The BattlegroundMgr is using the copy constructor for Battlegrounds in CreateNewBattleground. I think that's why I'm unable to have a vector of unique_ptr's?
```
error C2280: 'std::unique_ptr<WorldPackets::Battleground::BattlegroundPlayerPosition,std::default_delete<WorldPackets::Battleground::BattlegroundPlayerPosition>>::unique_ptr(const std::unique_ptr<WorldPackets::Battleground::BattlegroundPlayerPosition,std::default_delete<WorldPackets::Battleground::BattlegroundPlayerPosition>> &)': attempting to reference a deleted function (compiling source file H:\Trinity\source\src\server\game\Battlegrounds\BattlegroundMgr.cpp)
````

Shauren commented:
Convert it back to value and also:
* Declare copy constructor in Battleground class
* Create "implementation" in Battleground.cpp `Battleground::Battleground(Battleground const&) = default;`

Golrag committed:
Revert "Store positions as a shared_ptr"

This reverts commit 6e4a1b3226132838aa51eab6e4e6a91f36003044.

# Conflicts:
#	src/server/game/Spells/Auras/SpellAuraEffects.cpp

Golrag committed:
Define copy ctor

Golrag committed:
Clear container on reset

Shauren commented:
`_Copy_assign` errors in appveyor, add `Battleground& operator=(Battleground const&) = delete;`?

Golrag committed:
Delete  operator=(Battleground const&) function

Golrag commented:
That seems to work :)

Shauren commented on src/server/game/Battlegrounds/Battleground.cpp:
```diff
@@ -1669,6 +1684,21 @@ void Battleground::PSendMessageToAll(uint32 entry, ChatMsg msgType, Player const
+
+void Battleground::RemovePlayerPosition(ObjectGuid guid)
+{
+    auto itr = std::remove_if(_playerPositions.begin(), _playerPositions.end(), [guid](WorldPackets::Battleground::BattlegroundPlayerPosition const info)
```
add reference to lambda argument type

Shauren commented on src/server/game/Battlegrounds/Battleground.cpp:
```diff
@@ -1669,6 +1684,21 @@ void Battleground::PSendMessageToAll(uint32 entry, ChatMsg msgType, Player const
     va_end(ap);
 }
 
+void Battleground::AddPlayerPosition(WorldPackets::Battleground::BattlegroundPlayerPosition const position)
```
add reference to arg

Golrag committed:
Add references

Golrag commented:
Warsong Scramble is up now and I wonder how it behaves on retail. Does the last one that picked up the flags that is put in that arenaslot? Does it update automatically if that same player droppes the flag? Not sure how that all works

Side note: Warsong Scramble TLDR: 3 flags in each base, 10 points max

Golrag commented:
Okay, no idea about packets but I did it to get some quick answers:
- It updates automatically
- Its not the last one that picks it up or the first one that shows up. It can overwrite current ArenaSlot. (ex, i had flag, then druid, druid showed up on arena slot. Swapped it around then).

Shauren commented:
It just does this
```
ServerToClient: SMSG_BATTLEGROUND_PLAYER_POSITIONS (0x2928) Length: 80 ConnIdx: 1 Time: 02/07/2021 17:10:40.454 Number: 42649
BattlegroundPlayerPositionCount: 4
[0] Guid: Full: 1 Player
[0] Pos: X: 1130.923 Y: 1448.4581
[0] IconID: 1
[0] ArenaSlot: 3
[1] Guid: Full: 2 Player
[1] Pos: X: 1058.4755 Y: 1458.3544
[1] IconID: 1
[1] ArenaSlot: 3
[2] Guid: Full: 3 Player
[2] Pos: X: 1075.4467 Y: 1460.2341
[2] IconID: 1
[2] ArenaSlot: 3
[3] Guid: Full: 4 Player
[3] Pos: X: 1243.1815 Y: 1439.6768
[3] IconID: 2
[3] ArenaSlot: 2
```

Golrag commented:
Okay, well I think the code in this PR already supports this then. Maybe the client chooses the "correct" target

Shauren commented:
Before I merge - what happens with this PR if a player with one of these auras disconnects and reconnects?

Golrag commented:
I think that should trigger an aura remove, right?
Not sure what happens on retail. I think the character stays in the world for a while but unsure if the flag drops. Will test what happens in the PR

Shauren commented:
What I want to know is whether that creates duplicate entries for same guid in _playerPositions or not

Golrag commented:
GetBattleground is not nullptr and thus the player gets removed from the vector. BUT I noticed something that might be an issue.
I wait to log in till my character is removed from world, then login -> I get the aura back. Could this lead to possible other issues? Thinking about having multiple flag carriers suddenly. Might just be me using .cast instead of clicking the flag. Think its out of scope of this PR tho

Golrag commented:
And... I found a way to make battleground return nullptr: teleport the player while he is offline and carrying a flag

Golrag commented:
So this might happen when a player gets teleported out (hearthstone/teleport from gm). Which calls Battleground::RemovePlayerAtLeave which removes the battleground reference from player.

I have a fix for that one

Golrag committed:
Use BattlegroundMap to get battleground pointer
